File: 07--camel-cards/a.py
#!/usr/bin/env python3
'''
Usage:
    ./a.py PATH_TO_INPUT_FILE
'''

import sys
from collections import namedtuple, Counter
from enum import Enum, auto


Hand = namedtuple('Hand', 'cards bid')


class HandType(Enum):
    # From weakest to strongest
    HIGH_CARD = auto()
    ONE_PAIR = auto()
    TWO_PAIR = auto()
    THREE_OF_A_KIND = auto()
    FULL_HOUSE = auto()
    FOUR_OF_A_KIND = auto()
    FIVE_OF_A_KIND = auto()


def main():
    hands = []
    for line in open(sys.argv[1]).read().splitlines():
        cards, bid = line.split()
        bid = int(bid)
        hands.append(Hand(cards, bid))

    sorted_hands = sorted(
        hands,
        key = lambda hand: (
            get_hand_type(hand.cards).value,
            [get_card_value(card) for card in hand.cards]
        )
    )

    answer = sum(
        rank * hand.bid
        for rank, hand in enumerate(sorted_hands, start=1)
    )
    print(answer)


def get_hand_type(cards):
    counts = sorted(Counter(cards).values(), reverse=True)
    if counts == [5]:
        return HandType.FIVE_OF_A_KIND
    elif counts == [4, 1]:
        return HandType.FOUR_OF_A_KIND
    elif counts == [3, 2]:
        return HandType.FULL_HOUSE
    elif counts == [3, 1, 1]:
        return HandType.THREE_OF_A_KIND
    elif counts == [2, 2, 1]:
        return HandType.TWO_PAIR
    elif counts == [2, 1, 1, 1]:
        return HandType.ONE_PAIR
    elif counts == [1, 1, 1, 1, 1]:
        return HandType.HIGH_CARD
    else:
        raise Exception('unreachable case')


def get_card_value(card):
    try:
        return int(card)
    except ValueError:
        return {
            'A': 14,
            'K': 13,
            'Q': 12,
            'J': 11,
            'T': 10,
        }[card]


if __name__ == '__main__':
    main()

File: 07--camel-cards/a.hs
#!/usr/bin/env runhaskell
-- Usage:
--   ./a.hs PATH_TO_INPUT_FILE

import Data.Function ((&))
import Data.List (sort, sortOn, group)
import System.Environment (getArgs)


data Hand = Hand
  { cards :: String
  , bid :: Int
  } deriving (Show)

-- Could be just Int, but this makes things easier to debug, since you can
-- easily print a HandType and see what it is
type HandType = (Int, String)

fiveOfAKind = (7, "fiveOfAKind") :: HandType
fourOfAKind = (6, "fourOfAKind")
fullHouse = (5, "fullHouse")
threeOfAKind = (4, "threeOfAKind")
twoPair = (3, "twoPair")
onePair = (2, "onePair")
highCard = (1, "highCard")


main = do
  [inputPath] <- getArgs
  text <- readFile inputPath
  print $ solve text

solve text =
  zip [1..] hands
  & map (\(rank, hand) -> rank * (bid hand))
  & sum
  where
    hands =
      text
      & lines
      & map words
      & map
          (\[myCards, bidString] ->
            Hand myCards (read bidString)
          )
      & sortOn
          (\hand ->
            (
              getHandType $ cards hand,
              map getCardValue $ cards hand
            )
          )

getHandType myCards
  | counts == [5]
    = fiveOfAKind
  | counts == [4, 1]
    = fourOfAKind
  | counts == [3, 2]
    = fullHouse
  | counts == [3, 1, 1]
    = threeOfAKind
  | counts == [2, 2, 1]
    = twoPair
  | counts == [2, 1, 1, 1]
    = onePair
  | counts == [1, 1, 1, 1, 1]
    = highCard
  where
    counts =
      myCards
      & sort
      & group
      & map length
      & sort
      & reverse

getCardValue card
  | card == 'A' = 14
  | card == 'K' = 13
  | card == 'Q' = 12
  | card == 'J' = 11
  | card == 'T' = 10
  | otherwise   = read [card]
